runtime/common/src/assigned_slots.rs

```diff
@@ -0,0 +1,1086 @@
+					{
+						// Active slot lease
+						active_temp_slots += 1;
+					}
+					Some(last_lease)
+						// Slot w/ past lease, only consider it every other slot lease period (times period_count)
+						if last_lease.saturating_add(slot.period_count.saturating_mul(2u32.into())) <= lease_period_index => {
+							pending_temp_slots.push((para, slot));
+					},
+					None if slot.period_begin <= lease_period_index => {
+						// Slot hasn't had a lease yet
+						pending_temp_slots.insert(0, (para, slot));
+					},
+					_ => {
+						// Slot not being considered for this lease period (will be for a subsequent one)
+					},
+				}
+			}
+		});
+
+		if active_temp_slots < T::MaxTemporarySlotPerLeasePeriod::get() &&
+			!pending_temp_slots.is_empty()
+		{
+			// Sort by lease_count, favoring slots that had no or less turns first
+			// (then by last_lease index, and then Para ID)
+			pending_temp_slots.sort_by(|a, b| {
+				a.1.lease_count
+					.cmp(&b.1.lease_count)
+					.then_with(|| a.1.last_lease.cmp(&b.1.last_lease))
+					.then_with(|| a.0.cmp(&b.0))
+			});
+
+			let slots_to_be_upgraded = pending_temp_slots
+				.iter()
+				.take((T::MaxTemporarySlotPerLeasePeriod::get() - active_temp_slots) as usize)
+				.collect::<Vec<_>>();
+
+			for (id, temp_slot) in slots_to_be_upgraded.iter() {
+				TemporarySlots::<T>::try_mutate::<_, _, Error<T>, _>(id, |s| {
+					// Configure temp slot lease
+					Self::configure_slot_lease(
+						*id,
+						temp_slot.manager.clone(),
+						lease_period_index,
+						temp_slot.period_count,
+					)
+					.map_err(|_| Error::<T>::CannotUpgrade)?;
+
+					// Update temp slot lease info in storage
+					*s = Some(ParachainTemporarySlot {
+						manager: temp_slot.manager.clone(),
+						period_begin: temp_slot.period_begin,
+						period_count: temp_slot.period_count,
+						last_lease: Some(lease_period_index),
+						lease_count: temp_slot.lease_count + 1,
+					});
+
+					Ok(())
+				})?;
+			}
+		}
+		Ok(())
+	}
+
+	fn clear_slot_leases(origin: OriginFor<T>, id: ParaId) -> DispatchResult {
+		Slots::<T>::clear_all_leases(origin, id)
+	}
+
+	fn configure_slot_lease(
+		para: ParaId,
+		manager: T::AccountId,
+		lease_period: LeasePeriodOf<T>,
+		lease_duration: LeasePeriodOf<T>,
+	) -> Result<(), LeaseError> {
+		T::Leaser::lease_out(para, &manager, BalanceOf::<T>::zero(), lease_period, lease_duration)
+	}
+
+	fn has_permanent_slot(id: ParaId) -> bool {
+		PermanentSlots::<T>::contains_key(id)
+	}
+
+	fn has_temporary_slot(id: ParaId) -> bool {
+		TemporarySlots::<T>::contains_key(id)
+	}
+
+	fn lease_period_index() -> LeasePeriodOf<T> {
+		T::Leaser::lease_period_index()
+	}
+
+	fn lease_period() -> LeasePeriodOf<T> {
+		T::Leaser::lease_period()
+	}
+
+	fn manage_lease_period_start(lease_period_index: LeasePeriodOf<T>) -> Weight {
+		// Note: leases that have ended in previous lease period,
+		// should have been cleaned in slots pallet.
+		let _ = Self::allocate_temporary_slot_leases(lease_period_index);
```

Comment:
This drops any errors.
I would add logging, e.g. sth lik:
```suggestion
		let res = Self::allocate_temporary_slot_leases(lease_period_index);
		res.map_err(|e| log::error!(target: "assigned_slots", "Allocating slots failed for lease period {}, with: {}", lease_period_index, e));
```

Reply:
Good point, reworked it.